test/e2e/multicast_test.go

```diff
@@ -0,0 +1,295 @@
+			if len(mrouteResult) == 0 {
+				return false, nil
+			}
+		} else {
+			if len(mrouteResult) != 0 {
+				return false, nil
+			}
+		}
+		// Check whether multicast interfaces can receive multicast traffic in the server side.
+		// The check is needed for verifying external interfaces acting as multicast interfaces are able to forward multicast traffic.
+		// If multicast traffic is sent from non-HostNetwork pods, all multicast interfaces in senders should receive multicast traffic.
+		for _, multicastInterface := range multicastInterfaces[mc.senderConfig.idx] {
+			tcpdumpReceiveMulticastCommand := []string{"/bin/sh", "-c", fmt.Sprintf("timeout 5s tcpdump -q -i %s -c 1 host %s", multicastInterface, mc.group.String())}
+			_, stderr, err := data.runCommandFromPod(testNamespace, tcpdumpName, tcpdumpContainerName, tcpdumpReceiveMulticastCommand)
+			if err != nil {
+				return false, err
+			}
+			if mc.senderConfig.isHostNetwork && multicastInterface != transportInterface {
+				if !strings.Contains(stderr, "0 packets captured") {
+					return false, nil
+				}
+			} else {
+				if !strings.Contains(stderr, "1 packet captured") {
+					return false, nil
+				}
+			}
+		}
+		// Check inbound multicast route and whether multicast interfaces has joined the multicast group.
+		for _, receiver := range mc.receiverConfigs {
+			for _, receiverMulticastInterface := range multicastInterfaces[receiver.idx] {
+				_, mRouteResult, _, err := RunCommandOnNode(nodeName(receiver.idx), fmt.Sprintf("ip mroute show to %s iif %s ", mc.group.String(), receiverMulticastInterface))
+				if err != nil {
+					return false, err
+				}
+				// If multicast traffic is sent from non-HostNetwork pods and senders-receivers are located in different nodes,
+				// the receivers should configure corresponding inbound multicast routes.
+				if mc.senderConfig.idx != receiver.idx && !receiver.isHostNetwork {
+					if len(mRouteResult) == 0 {
+						return false, nil
+					}
+				} else {
+					if len(mRouteResult) != 0 {
+						return false, nil
+					}
+				}
+				_, mAddrResult, _, err := RunCommandOnNode(nodeName(receiver.idx), fmt.Sprintf("ip maddr show %s | grep %s", receiverMulticastInterface, mc.group.String()))
+				if err != nil {
+					return false, err
+				}
+				// The receivers should also join multicast group.
+				// Note that in HostNetwork mode, the "join multicast" action is taken by mcjoin,
+				// which will not persist after mcjoin exits.
+				if !receiver.isHostNetwork {
+					if len(mAddrResult) == 0 {
+						return false, nil
+					}
+				} else {
+					if len(mAddrResult) != 0 {
+						return false, nil
+					}
+				}
+			}
+		}
+		return true, nil
+	}); err != nil {
+		t.Fatalf("Error when waiting for multicast routes and stats: %v", err)
+	}
+	wg.Wait()
+}
+
+// computeMulticastInterfaces computes multicastInterfaces for each node.
+// It return map[int][]string with its key as node index and value multicastInterfaces.
```

Comment:
Since the index is continuous, why don't return slice directly, which is more natural for linear data.

Reply:
updated